Feat private range proof

thomas-waite opened the pull request:
**This PR is largely similar to a previous PR that was reviewed and approved by Zac (https://github.com/AztecProtocol/AZTEC/pull/91) and merged. This PR has migrated the previously merged proof to the new construction.**

## Summary
This PR implements a zero-knowledge proof that a confidential digital asset has a value greater or less than another confidential digital asset. This is a private range range proof.

This PR includes: JavaScript proof construction and verification code, ABI input encoding, smart contract zk proof validator, smart contract output ABI encoding and tests. 

## Types of changes
**New feature**
This PR implements a new feature - a privateRange proof. There are no breaking changes.

## Future work/next steps
The `utilityNote` should be abstracted away from the proof construction interface, it is currently exposed. It serves as a 'behind the scenses' helper note used to construct a proof statement.

Abstracting the `utilityNote` away such that it would be constructed inside the proof, would naively require making the constructor asynchronous - will consider the best way to avoid this.

thomas-waite committed:
refactor: metacommit - work in progress

can not yet get last blinding factor to match across both proof construction and verification. Same
problem with both the old proof and new proof format

thomas-waite committed:
feat(aztec.js): implement privateRange proof JS code
JS code for proof construction and verification

thomas-waite committed:
feat(dev-utils): add PRIVATE_RANGE_PROOF ID

thomas-waite committed:
test(aztec.js): implement tests for JS private range verifier

thomas-waite committed:
feat(protocol): implement privateRange proof smart contract verifier
implement verifier and ABI encoder
add interface

thomas-waite committed:
test(protocol): tests for privateRange verifier contracts

thomas-waite committed:
refactor: eslint fixes, remove unnecessary comments

thomas-waite committed:
docs(aztec.js): add JSDoc for proof construction method

thomas-waite committed:
refactor(aztec.js): remmove oldPrivateRange debugging files

zac-williamson commented on packages/aztec.js/test/proof/privateRange/index.js:
```diff
@@ -0,0 +1,48 @@
+        });
+    });
+
+    describe('Failure States', () => {
```
We should test that the proof fails if the input note is not within the desired range.

This should fail in two locations - when constructing the proof, we should return an error state, and we should also test that verifying such a proof fails

zac-williamson replied:
Oh, I was looking at the wrong file! We should still not construct a proof over bad inputs though, as this could cause misuse of our API

thomas-waite replied:
Good point. Have implemented this check for the privateRange proof construction, but ideally we should also do this for all proofs

zac-williamson commented on packages/protocol/contracts/ACE/validators/privateRange/PrivateRange.sol:
```diff
@@ -0,0 +1,346 @@
+                            result,
+                            and(
+                                and(
+                                    staticcall(gas, 6, 0x1a0, 0x80, 0x1e0, 0x40),
```
We have 3 notes in this proof, but we are only checking the pairing relationship on 2 notes

In this section, we need to update the memory location that we reference when we call the ecAdd precompile, so that we store the 3 notes linearly in memory

zac-williamson commented on packages/protocol/test/ACE/validators/privateRange/index.js:
```diff
@@ -0,0 +1,324 @@
+            );
+        });
+
+        it('should fail if blindingFactors NOT integrated into challenge variable', async () => {
```
I think the following test would be useful to add, to check that we're validating our bilinear pairing correctly:

Create a situation where the input/output/utility note values satisfy the private range proof, but the utility note is generated over a different setup key than the other notes (i.e. `y` is not the same, so the bilinear pairing test should fail). 

We should repeat the above test for permutations of all 3 notes (so one where the input note is over the wrong setup key, and one for the output note)

thomas-waite replied:
Very good point. As discussed, we want to implement this across all proofs. Will add the task to our scheduling tool and break it out seperately from this PR

zac-williamson reviewed (changes requested):
This is great work. I like how clean the proof construction interface is and how methodical the tests are. The API has come a long way!

I couldn't find any problems with the cryptography, but I think there are a couple of extra tests that we would benefit from, to validate that our bilinear pairing check is preventing rogue commitments (created from a different trusted setup) from being added into the proof.

thomas-waite committed:
feat(aztec.js): add proof construction inputs check
check that the supplied notes satisfy the privateRange balancing relationship, throw if not. Stops proofs being created by the client that will be rejected by the verifier
add failure test to confirm behaviour

thomas-waite committed:
refactor(dev-utils): add BALANCING_RELATIONSHIP_NOT_SATISFIED error code

prettier fix for privateRange proof construction
